main/src/ca/uwaterloo/flix/language/phase/unification/FastBoolUnification.scala

```diff
@@ -0,0 +1,1694 @@
+      *
+      * Note: Does not make sense to call before the equation system has been fully solved.
+      *
+      * Throws a [[ConflictException]] if an equation is not solved by the current substitution.
+      */
+    private def verifySolution(): Unit = {
+      verify(currentSubst, l)
+    }
+
+    /**
+      * Checks that current substitution has not grown too large according to the defined threshold.
+      */
+    private def verifySolutionSize(): Unit = {
+      val size = currentSubst.size
+      if (size > Threshold) {
+        throw TooComplexException(size)
+      }
+    }
+
+    private def printEquations(): Unit = {
+      debugln(s"Equations (${currentEqns.size}):")
+      debugln(format(currentEqns))
+    }
+
+    private def printSubstitution(): Unit = {
+      debugln(s"Substitution (${currentSubst.numberOfBindings}):")
+      debugln(currentSubst.toString)
+    }
+
+    private def debugln(): Unit = debugln("")
+
+    private def debugln(s: String): Unit = {
+      if (Debugging) {
+        Console.println(s)
+      }
+    }
+
+  }
+
+  /**
+    * Returns a list of non-trivial unification equations computed from the given list `l`.
+    *
+    * Throws a [[ConflictException]] if an unsolvable equation is encountered.
+    *
+    * A trivial equation is one of:
+    * -     true ~ true
+    * -    false ~ false
+    * -        c ~ c        (same constant)
+    * -        x ~ x        (same variable)
+    *
+    *
+    * An unsolvable (conflicted) equation is one of:
+    *
+    * -      true ~ false   (and mirrored)
+    * -       c_i ~ c_j     (different constants)
+    * -        c ~ true     (and mirrored)
+    * -        c ~ false    (and mirrored)
+    */
+  private def simplify(l: List[Equation]): List[Equation] = l match {
+    case Nil => Nil
+    case Equation(t1, t2, loc) :: es => (t1, t2) match {
+      // Trivial equations: skip them.
+      case (Term.True, Term.True) => simplify(es)
+      case (Term.False, Term.False) => simplify(es)
+      case (Term.Cst(c1), Term.Cst(c2)) if c1 == c2 => simplify(es)
+      case (Term.Var(x1), Term.Var(x2)) if x1 == x2 => simplify(es)
+
+      // Unsolvable (conflicted) equations: raise an exception.
+      case (Term.True, Term.False) => throw ConflictException(t1, t2, loc)
+      case (Term.False, Term.True) => throw ConflictException(t1, t2, loc)
+      case (Term.Cst(c1), Term.Cst(c2)) if c1 != c2 => throw ConflictException(t1, t2, loc)
+      // Note: A constraint with two different variables is of course solvable!
+      case (Term.Cst(_), Term.True) => throw ConflictException(t1, t2, loc)
+      case (Term.Cst(_), Term.False) => throw ConflictException(t1, t2, loc)
+      case (Term.True, Term.Cst(_)) => throw ConflictException(t1, t2, loc)
+      case (Term.False, Term.Cst(_)) => throw ConflictException(t1, t2, loc)
+
+      // Non-trivial and non-conflicted equation: keep it.
+      case _ => Equation(t1, t2, loc) :: simplify(es)
```

Comment:
make it tail-recursive? or is the number of constraints small per function?

Reply:
Very small: linear in code size. If functions are big, many other phases of the compiler will blow up before.